external/avro/src/main/scala/org/apache/spark/sql/avro/SparkAvroDatumReader.scala

```diff
@@ -452,71 +452,73 @@ class SparkAvroDatumReader[T](
-        array.sizeHint((array.length + length).toInt)
+        array = arrayExpander(arrayUpdater, length) // if (length == 0) 0 else 1)
       }
 
-      updater.set(ordinal, toArrayConverter())
-      array.clear()
+      updater.set(ordinal, array)
     }
   }
 
   private[this] def getMapReader(avroType: Schema,
       keyType: DataType,
       valueType: DataType,
-      path: List[String],
-      reuseObj: Boolean
+      path: List[String]
   ): (CatalystDataUpdater, Int, ResolvingDecoder) => Unit = {
     val keyReader = newReader(SchemaBuilder.builder().stringType(), StringType, path, false)
-    val keyArray = new ArrayBuffer[Any]
-    val keyArrayUpdater = new ArrayBufferUpdater(keyArray)
-    val toKeyArrayConverter = getToArrayDataConverter(keyType, keyArray)
+    val keyArrayCreator = getArrayDataCreator(keyType)
+    val keyArrayExpander = getArrayDataExpander(keyType)
 
     val valueReader = newReader(avroType.getValueType, valueType, path, false)
-    val valueArray = new ArrayBuffer[Any]
-    val valueArrayUpdater = new ArrayBufferUpdater(valueArray)
-    val toValueArrayConverter = getToArrayDataConverter(valueType, valueArray)
+    val valueArrayCreator = getArrayDataCreator(valueType)
+    val valueArrayExpander = getArrayDataExpander(valueType)
 
     (updater, ordinal, in) => {
       var length = in.readMapStart()
-      keyArray.sizeHint(length.toInt)
-      valueArray.sizeHint(length.toInt)
+      var keyArray = keyArrayCreator(length) // if (length == 0) 0 else 1)
+      var valueArray = valueArrayCreator(length) // if (length == 0) 0 else 1)
+      val keyArrayUpdater = new ArrayDataUpdater(keyArray)
+      val valueArrayUpdater = new ArrayDataUpdater(valueArray)
 
+      var base: Int = 0
       while (length > 0) {
         var i = 0
         while (i < length) {
-          keyReader(keyArrayUpdater, i, in)
-          valueReader(valueArrayUpdater, i, in)
+          keyReader(keyArrayUpdater, base + i, in)
+          valueReader(valueArrayUpdater, base + i, in)
           i += 1
+          // keyArray = keyArrayExpander(keyArrayUpdater, if (i == length) 0 else 1)
+          // valueArray = valueArrayExpander(valueArrayUpdater, if (i == length) 0 else 1)
         }
+        base += length.toInt
         length = in.mapNext()
-        keyArray.sizeHint((keyArray.length + length).toInt)
-        valueArray.sizeHint((valueArray.length + length).toInt)
+        keyArray = keyArrayExpander(keyArrayUpdater, length) // if (length == 0) 0 else 1)
+        valueArray = valueArrayExpander(valueArrayUpdater, length) // if (length == 0) 0 else 1)
       }
 
       // The Avro map will never have null or duplicated map keys, it's safe to create a
       // ArrayBasedMapData directly here.
-      updater.set(ordinal, new ArrayBasedMapData(toKeyArrayConverter(), toValueArrayConverter()))
-      keyArray.clear()
-      valueArray.clear()
+      updater.set(ordinal, new ArrayBasedMapData(keyArray, valueArray))
     }
   }
```

Comment:
Changes to reading maps is the same as the ones to reading arrays https://github.com/apache/spark/pull/29354/commits/2c62ac9ae960582058e96860002f7768eebb95f2#r465994403